### PR TITLE
Typescript more r misc

### PR DESCRIPTION
**Pull request by johnjeng**
# Description
Typescript basic parts of r_misc.
Move `r_misc` and tests into its own folder.
Use module resolution (`r_misc.cjsx` -> `r_misc/index.cjsx`)

# Testing Steps
1. Run `jest`
1. 
1.

# Relevant Issues
Follow up on #3333 

### [Checklist](https://github.com/sagemathinc/cocalc/wiki/PR-Checklist):
- [x] No debugging console.log messages.
- [x] All new code is actually used.
- [x] Non-obvious code has some sort of comments.

Front end:
- [x] Restart at least one project and check its `~/.smc/local_hub/local_hub.log`
- [x] Completely restart Webpack with `./w` in `/src`
- [x] Screenshots if relevant.

### CODE REVIEW

File: src/smc-webapp/r_misc/selector-input.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    | string[]
+    | { value: string; display: JSX.Element }[]
+    | { [keys: string]: JSX.Element };
+  disabled?: boolean;
+  selected?: string;
+  on_change?: (selected: string) => void;
+}
+
+// If the first element is a string, we assume the rest to be a string
+function isStringArrayHeuristic(a: any): a is string[] {
```
Comment (williamstein) [inline src/smc-webapp/r_misc/selector-input.tsx]: What does `a is string[]` mean for the return type?  Shouldn't the return type be `boolean`???

I'm not doubting this is valid typescript; it's just that I don't understand it....

File: src/smc-webapp/r_misc/close-x.tsx [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import * as React from "react";
+import { Icon } from "./icon";
+
+const closex_style: React.CSSProperties = {
```
Comment (williamstein) [inline src/smc-webapp/r_misc/close-x.tsx]: Oh, so *this* (use `React.CSSProperties`) is how to declare objects and using them for css.  I've been confused about this.    I should have asked you in chat.

File: src/smc-webapp/r_misc/selector-input.tsx (hunk 1 above)
Comment (johnjeng) [reply src/smc-webapp/r_misc/selector-input.tsx]: This is the syntax to tell Typescript that is function is a typechecker. eg
```
// foo: string[] | number;
let length: number = 0;
if (isStringArrayHeuristic(foo)) { // Behaves as if we could do typeof foo == string[]
  length: number = foo.length;  // <-- TS knows foo is a string array so has property length 
} else {
  length = foo; // <-- TS knows foo is a number
}
```

Comment (williamstein) [reply src/smc-webapp/r_misc/selector-input.tsx]: (and this code is all merged into master and live on prod already, so this is purely for my education.)

Comment (johnjeng) [reply src/smc-webapp/r_misc/selector-input.tsx]: In the documentation is called a "User-Defined Type Guards".

It's on this page https://www.typescriptlang.org/docs/handbook/advanced-types.html

**Pull request merged**